Ufactory Lite6 gripper model

eufrizz opened the pull request:
I modelled the gripper for the Ufactory Lite6 for my own use so I thought I'd share it in case it's useful for anyone else! (Although I suspect it doesn't have a large number of users, and the gripper itself is somewhat rubbish 😅)

The stl files were created in Onshape from the .step files released by the manufacturer. I added two new models, one with the grippers internally rotated, and one with them externally rotated. More detail in the readme.

I wasn't entirely sure how best to model the solver characteristics. The ones chosen are taken from the ACT repo for the Aloha gripper, I couldn't get the default values to work nicely with picking up a box (meshes would intersect), so happy for someone who knows more to improve on this.
Similarly, the friction characteristics are an approximation for the rubberised pads on the gripper fingers.

kevinzakka commented on ufactory_lite6/README.md:
```diff
@@ -23,6 +23,21 @@ description](https://github.com/xArm-Developer/xarm_ros2/tree/master/xarm_descri
 5. Added `scene.xml` which includes the robot, with a textured groundplane, skybox, and haze.
 
+### Gripper
+
```
> The fingers can be affixed rotated internally

Typo?

eufrizz replied:
In the sense of "affixed with an internal rotation" but I'll rewrite it all to be clearer

kevinzakka replied:
Gotcha!

kevinzakka commented on ufactory_lite6/README.md:
```diff
@@ -23,6 +23,21 @@ description](https://github.com/xArm-Developer/xarm_ros2/tree/master/xarm_descri
+
+The URDF only contains a gripper model for the Lite 6 with the fingers fused to the gripper body. To separate the fingers, the .step file of the fused body from the [Ufactory website](https://usa.ufactory.cc/download-lite6-robot) was loaded into Onshape ([file here](https://cad.onshape.com/documents/f60aac1c8ff6af8f490dc855/w/5c0df4bc7414802fc89a514e/e/7dc41825dd66894c14b085ca?renderMode=0&uiStat ...
+
+Because they are actuated via an air compressor, a single force actuator was used in the model, with an equality constraint to mimic the gearing mechanism that keeps them equidistant from the centre.
```
> single force actuator

Can you specify that you used a `motor`?

kevinzakka commented on ufactory_lite6/README.md:
```diff
@@ -23,6 +23,21 @@ description](https://github.com/xArm-Developer/xarm_ros2/tree/master/xarm_descri
+
+Because they are actuated via an air compressor, a single force actuator was used in the model, with an equality constraint to mimic the gearing mechanism that keeps them equidistant from the centre.
+
+The inertial parameters were estimated from Onshape, in lieu of having any more detailed data, by assuming the body is composed from uniformly dense material chosen to give the correct the mass.
```
Nit: can you simplify this sentence? It's a bit convoluted!

kevinzakka reviewed (commented):
First, thank you so much for contributing to Menagerie, we really appreciate it!

I have two major questions before embarking on a review of the actual XML files:

1) I don't fully understand the difference between the external and internal gripper variants. It looks like only the internal one is needed? Additionally, "internal" vs "external" seems like the wrong word to use and can cause confusion.
2) This one might require @yuvaltassa 's input, but with the introduction of the new `MjSpec` API, do we need to have different variants of the arm (arm alone, arm + gripper, gripper alone) or should we switch to a cleaner and more maintainable (standalone arm, standalone gripper) and attach together via via `MjSpec`?

eufrizz commented:
@kevinzakka thanks for the review!
1. Open to suggestions for better terminology, I think "internal" and "external" made sense to me from the terminology used for shoulders (internal vs external rotation, see attached photo). I've changed it to say internal and external _facing_ in the readme, this is probably clearer than "rotation".
Because the range of motion of the gripper is so short, the maximum distance between the grippers is only 15mm in the internal configuration so you can only pick up narrow objects. In the external configuration, you get 38mm. Thus it is useful to have both.

2. I was not aware of mjSpec, this is great, keen to clean up the mess of all these different XMLs! I'll have a play around with it in the next couple days.

![](https://s-media-cache-ak0.pinimg.com/236x/7d/5c/a3/7d5ca3350346974feb37bd33966557ec.jpg)

yuvaltassa commented:
You'll want to look at the new [attach element](https://mujoco.readthedocs.io/en/latest/XMLreference.html#body-attach)

kevinzakka commented on ufactory_lite6/lite6_gripper_external.xml:
```diff
@@ -0,0 +1,152 @@
+    <default class="lite6">
+      <geom type="mesh"/>
+      <joint axis="0 0 1" damping="1" armature="0.1"/>
+      <general gaintype="fixed" biastype="affine" gainprm="2000" biasprm="0 -2000 -200"/>
```
Do you want to switch to position actuators which now support damping? This can be converted to:

`<position kp="2000" kv="200"/>`

kevinzakka replied:
FYI, if you update this, make sure you change the minimum required version for MuJoCo at the top of the README.

eufrizz replied:
Done. Updated Mujoco version to 3.1.0 when kv was introduced for position actuators as per [changelog](https://mujoco.readthedocs.io/en/latest/changelog.html#version-3-1-0-december-12-2023)

kevinzakka commented on ufactory_lite6/lite6_gripper_internal.xml:
```diff
@@ -0,0 +1,143 @@
+    <default class="lite6">
+      <geom type="mesh"/>
+      <joint axis="0 0 1" damping="1" armature="0.1"/>
+      <general gaintype="fixed" biastype="affine" gainprm="2000" biasprm="0 -2000 -200"/>
```
Same comment here.

kevinzakka reviewed (changes requested):
Thanks @eufrizz, I think we're almost there. Just left a comment to improve the readability of the XML by switching from general to the equivalent position actuator.

eufrizz commented:
@kevinzakka thanks. Updated the original (no gripper) model with position actuators too. I also added names to the collision geoms (helps when checking for self intersections etc.) and updated this in the original model for consistency.

eufrizz commented:
@kevinzakka @yuvaltassa I had a play around with the new MjSpec/attach functionality. 
I found that the Python API didn't work - `mujoco.MjSpec().from_file("...scene.xml")` returns None (and crashes if the wrong filename is provided).
But I got it going with the C++ API (just compiled and visually inspected, didn't test too thoroughly). I found it necessary to attach a prefix and/or suffix at the `mis_attachBody` step to avoid a compilation error which seemed a little quirky.
I think it makes more sense to do as a separate PR later on perhaps when the functionality has matured a bit/when other models will follow this pattern too.
I'll happily submit a draft PR with the changes and a bit of demo code now though so you can take a look, and merge when you think it's appropriate.

kevinzakka commented:
Awesome job @eufrizz, will try to get this merged next week alongside the other PRs :)

eufrizz commented:
Also, still unsure on the chosen solver parameters for the gripper. Would love a double check on that.

kevinzakka commented:
I’ll tweak during the merge and let you know what we change!
